src/schemes/fluid/transport_velocity.jl

```diff
@@ -7,176 +7,117 @@ See [TVF](@ref transport_velocity_formulation) for more details of the method.
 
-function reset_callback_flag!(system::FluidSystem, ::TransportVelocityAdami)
-    system.cache.update_callback_used[] = false
+        system_coords = current_coordinates(u, system)
+        neighbor_coords = current_coordinates(u_neighbor, neighbor_system)
 
-    return system
-end
+        foreach_point_neighbor(system, neighbor_system, system_coords, neighbor_coords,
+                               semi;
+                               points=each_moving_particle(system)) do particle, neighbor,
+                                                                       pos_diff, distance
+            m_a = hydrodynamic_mass(neighbor_system, neighbor)
+            m_b = hydrodynamic_mass(neighbor_system, neighbor)
 
-function update_callback_used!(system::FluidSystem)
-    update_callback_used!(system, system.transport_velocity)
-end
+            grad_kernel = smoothing_kernel_grad(system, pos_diff, distance, particle)
 
-update_callback_used!(system, ::Nothing) = system
+            # The TVF is based on the assumption that the pressure gradient is only accurately
+            # computed when the particle distribution is isotropic.
+            # That means, the force contribution vanishes only if the particle distribution is
+            # isotropic AND the field being differentiated by the kernel gradient is spatially constant.
+            # So we must guarantee a constant field and therefore the reference density is used
+            # instead of the locally computed one.
+            # TODO:
+            # volume_a = particle_spacing(system, particle)^ndims(system)
+            # volume_b = particle_spacing(neighbor_system, neighbor)^ndims(neighbor_system)
+            volume_a = m_a / system.initial_condition.density[particle]
+            volume_b = m_b / neighbor_system.initial_condition.density[neighbor]
 
-function update_callback_used!(system, transport_velocity)
-    system.cache.update_callback_used[] = true
-end
+            volume_term = (volume_a^2 + volume_b^2) / m_a
 
-function check_tvf_configuration(system::FluidSystem, ::Nothing,
-                                 v, u, v_ode, u_ode, semi, t; update_from_callback=false)
-    return system
-end
+            delta_v_ = -volume_term * background_pressure * grad_kernel
 
-function check_tvf_configuration(system::FluidSystem, ::TransportVelocityAdami,
-                                 v, u, v_ode, u_ode, semi, t; update_from_callback=false)
-    # The `UpdateCallback` will set `system.cache.update_callback_used[]` to `true`
-    # in the initialization. However, other callbacks might update the system first, hence `update_from_callback`.
-    if !update_from_callback && !(system.cache.update_callback_used[])
-        throw(ArgumentError("`UpdateCallback` is required when using `TransportVelocityAdami`"))
-    end
+            # Write into the buffer
+            for i in eachindex(delta_v_)
+                @inbounds delta_v[i, particle] += delta_v_[i]
```

Comment:
This needs to contain the time step:
```suggestion
                @inbounds delta_v[i, particle] += time_step * delta_v_[i]
```
I hardcoded this for testing to produce the results in the PR description.